src/backend/distributed/executor/local_executor.c

```diff
@@ -468,49 +468,44 @@ ExtractLocalAndRemoteTasks(bool readOnly, List *taskList, List **localTaskList,
 		/* either the local or the remote should be non-nil */
 		Assert(!(localTaskPlacementList == NIL && remoteTaskPlacementList == NIL));
 
-		if (list_length(task->taskPlacementList) == 1)
+		if (list_length(localTaskPlacementList) > 0)
 		{
-			/*
-			 * At this point, the task has a single placement (e.g,. anchor shard
-			 * is distributed table's shard). So, it is either added to local or
-			 * remote taskList.
-			 */
-			if (localTaskPlacementList == NIL)
+			if (list_length(task->taskPlacementList) == 1)
 			{
-				*remoteTaskList = lappend(*remoteTaskList, task);
+				*localTaskList = lappend(*localTaskList, task);
 			}
 			else
 			{
-				*localTaskList = lappend(*localTaskList, task);
-			}
-		}
-		else
-		{
-			/*
-			 * At this point, we're dealing with reference tables or intermediate
-			 * results where the task has placements on both local and remote
-			 * nodes. We always prefer to use local placement, and require remote
-			 * placements only for modifications.
-			 */
-			task->partiallyLocalOrRemote = true;
+				/*
+				 * At this point, we're dealing with reference tables or intermediate
+				 * results where the task has placements on both local and remote
+				 * nodes. We always prefer to use local placement, and require remote
+				 * placements only for modifications.
+				 */
+				task->partiallyLocalOrRemote = true;
 
-			Task *localTask = copyObject(task);
+				Task *localTask = copyObject(task);
 
-			localTask->taskPlacementList = localTaskPlacementList;
-			*localTaskList = lappend(*localTaskList, localTask);
+				localTask->taskPlacementList = localTaskPlacementList;
+				*localTaskList = lappend(*localTaskList, localTask);
 
-			if (readOnly)
-			{
-				/* read-only tasks should only be executed on the local machine */
-			}
-			else
-			{
-				Task *remoteTask = copyObject(task);
-				remoteTask->taskPlacementList = remoteTaskPlacementList;
+				if (readOnly)
+				{
+					/* read-only tasks should only be executed on the local machine */
+				}
+				else
+				{
+					Task *remoteTask = copyObject(task);
+					remoteTask->taskPlacementList = remoteTaskPlacementList;
```

Comment:
(out of the changes done in this pr)
We can add such an assertion & comment here to make it more clear for readers that we do not create remote task with empty task placements

```suggestion
                                /* for non read-only tasks, task would definitely have remote placements */
                                Assert(remoteTaskPlacementList != NIL);
                                
				else
				{
					Task *remoteTask = copyObject(task);
					remoteTask->taskPlacementList = remoteTaskPlacementList;
```